feat: Gateway Intents article (#354)
discordjs/guide: guide/popular-topics/intents.md

# Gateway Intents

Gateway Intents were introduced to the library in v12 and allow you to pick-and-choose which events you wish your bot to receive and/or omit.
Intents are groups of pre-defined events that the discord.js client will conditionally subscribe to. For example, omitting the `GUILD_PRESENCES` intent would prevent the discord.js client from receiving any member-related presence events. Before this feature, the discord.js client would receive all events, thus consuming memory for data you may not be using.

:::warning
Intents are optional on the v6 gateway but will become mandatory in a future gateway version.
:::

<branch version="11.x">

Intents are not available in version 11, please update to version 12 of the library if you want to use gateway intents in your bot.

</branch>

<branch version="12.x">

## Enabling Intents

> You can choose which intents you'd like to receive as client options when instantiating your bot client.

Reviewer: Intents in discord.js are built on top of the BitField wrapper
At least a short introduction to computing custom intents should be included in this guide (does not need to be as in-depth as the permissions guide, of course) - the computing starting from <https://discord.js.org/#/docs/main/stable/class/Intents?scrollTo=s-ALL> preferably

Good intends to use for an example here are "DIRECT_MESSAGE_TYPING" and "GUILD_MESSAGE_TYPING" as a lot of bots don't have any typing functionality and this is a rather big impact we already recommended with disabling events for people wanting to cut memory usage down

  - `guildIntegrationsUpdate`
- `GUILD_WEBHOOKS` - Events from this intent include:
  - `webhookUpdate`
- `GUILD_INVITES` - Events from this intent include:
  - `inviteCreate`
  - `inviteDelete`
- `GUILD_VOICE_STATES` - Events from this intent include:
  - `voiceStateUpdate`
- `GUILD_PRESENCES`¹ - Events from this intent include (within guilds):
  - `presenceUpdate`
- `GUILD_MESSAGES` - Events from this intent include:
  - `message`
  - `messageUpdate`
  - `messageDelete`
- `GUILD_MESSAGE_REACTIONS` - Events from this intent include (within guilds):
  - `messageReactionAdd`
  - `messageReactionRemove`
  - `messageReactionRemoveAll`
  - `messageActionRemoveEmoji`
- `GUILD_MESSAGE_TYPING` - Events from this intent include (within guilds):
  - `typingStart`
- `DIRECT_MESSAGES` - Events from this intent include (within DMs):
  - `channelCreate`
  - `message`
  - `channelUpdate`
  - `channelDelete`
  - `channelPinsUpdate`
- `DIRECT_MESSAGE_REACTIONS` - Events from this intent include (within DMs):
  - `messageReactionAdd`
  - `messageReactionRemove`
  - `messageReactionRemoveAll`
  - `messageActionRemoveEmoji`
- `DIRECT_MESSAGE_TYPING` - Events from this intent include (within DMs):

>   - `typingStart`

Reviewer: This should not be documented in the guide, as it is directly taken from  <https://discordapp.com/developers/docs/topics/gateway#list-of-intents> it might as well just link there and explain what the headings mean and how to use them on an example

¹privileged intent (see [Privileged Intents](#privileged-intents))

:::danger
At the time of writing (2020/03/23), discord.js does not set default intents, which could potentially break things.
See [this](https://github.com/discordjs/discord.js/issues/3924) issue for more information.
:::

:::tip
Make sure you enable all intents you need for your use case! If you miss one, the corresponding events do not get emitted. For example, If you miss the `GUILDS` intent, your bot won't receive the `guildCreate` event, which is crucial if you implement per-guild command prefixes and create a new database entry when the bot is added to new guilds. If you miss the `GUILD_MESSAGES` intent, your client won't receive messages sent within guilds.
:::

:::tip
We strongly recommend you omit the `GUILD_PRESENCES` intent if you don't frequently access the presence data.
:::

```js
const { Client } = require('discord.js');
const client = new Client({ ws: { intents: ['GUILDS', 'GUILD_MESSAGES'] } });
```

> ### Privileged Intents

Reviewer: This section should mention (and maybe even suggest) <https://discord.js.org/#/docs/main/stable/class/Intents?scrollTo=s-NON_PRIVILEGED> as good fallback without needing to provide every single intent.

# Gateway Intents

> Gateway Intents were introduced to the library in v12 and allow you to pick-and-choose which events you wish your bot to recieve and/or omit.

Reviewer 2: ```suggestion
Gateway Intents were introduced to the library in v12 and allow you to pick-and-choose which events you wish your bot to receive and/or omit.
```

> Intents are groups of pre-defined events which the discord.js client will conditionally subscribe to. For example, omiting the `GUILD_PRESENCES` intent would prevent the discord.js client from recieving any member-related presence events. Prior to this feature, the discord.js client would recieve all events, thus consuming memory for data you may not use.

Reviewer 2: ```suggestion
Intents are groups of pre-defined events which the discord.js client will conditionally subscribe to. For example, omitting the `GUILD_PRESENCES` intent would prevent the discord.js client from receiving any member-related presence events. Prior to this feature, the discord.js client would receive all events, thus consuming memory for data you may not use.
```

# Gateway Intents

Gateway Intents were introduced to the library in v12 and allow you to pick-and-choose which events you wish your bot to recieve and/or omit.

> Intents are groups of pre-defined events which the discord.js client will conditionally subscribe to. For example, omiting the `GUILD_PRESENCES` intent would prevent the discord.js client from recieving any member-related presence events. Prior to this feature, the discord.js client would recieve all events, thus consuming memory for data you may not use.

Reviewer 3: ```suggestion
Intents are groups of pre-defined events which the discord.js client will conditionally subscribe to. For example, omiting the `GUILD_PRESENCES` intent would prevent the discord.js client from receiving any member-related presence events. Prior to this feature, the discord.js client would recieve all events, thus consuming memory for data you may not use.
```

Reviewer 2: This is already covered by my review, and you missed the other typos anyway

Reviewer 3: Woops, I overlooked that, I apologize.

- `GUILD_WEBHOOKS` - Events from this intent include:
  - `webhookUpdate`
- `GUILD_INVITES` - Events from this intent include:
  - `inviteCreate`
  - `inviteDelete`
- `GUILD_VOICE_STATES` - Events from this intent include:
  - `voiceStateUpdate`
- `GUILD_PRESENCES`¹ - Events from this intent include (within guilds):
  - `presenceUpdate`
- `GUILD_MESSAGES` - Events from this intent include:
  - `message`
  - `messageUpdate`
  - `messageDelete`
- `GUILD_MESSAGE_REACTIONS` - Events from this intent include (within guilds):
  - `messageReactionAdd`
  - `messageReactionRemove`
  - `messageReactionRemoveAll`
  - `messageActionRemoveEmoji`
- `GUILD_MESSAGE_TYPING` - Events from this intent include (within guilds):
  - `typingStart`
- `DIRECT_MESSAGES` - Events from this intent include (within DMs):
  - `channelCreate`
  - `message`
  - `channelUpdate`
  - `channelDelete`
  - `channelPinsUpdate`
- `DIRECT_MESSAGE_REACTIONS` - Events from this intent include (within DMs):
  - `messageReactionAdd`
  - `messageReactionRemove`
  - `messageReactionRemoveAll`
  - `messageActionRemoveEmoji`
- `DIRECT_MESSAGE_TYPING` - Events from this intent include (within DMs):
  - `typingStart`

> ¹privleged intent (see [Privleged Intents](#privileged-intents))

Reviewer 3: ```suggestion
privileged intent (see [Privileged Intents](#privileged-intents))
```
